### DISCUSSION TITLE
Improve defaultValue types and allow relationship to accept more than a function

### DISCUSSION THREAD

**Discussion by evelynhathaway** [payloadcms/payload#1720] [Roadmap]
# Documentation Issue

The [relationship field documentation](https://payloadcms.com/docs/fields/relationship#:~:text=defaultValue,default%20value.%20More) doesn't provide a description of the type or what data should be included for the `defaultValue` option.

Improving the TypeScript types for `CollectionConfig` and the documentation could help new users.

## Additional Details

I ran into errors while trying to add a default value to a relationship field

```
There were 1 errors validating your Payload config
Collection "posts" > Field "category" > "value" does not match any of the allowed types
```

I had to dive into the source code to find it only accepts a function

https://github.com/payloadcms/payload/blob/934b443b5b3b0f610767786f940175b9db0b2da7/src/fields/config/schema.ts#L313-L315

Before I found that, I tried using:
- `id` as a string
- `name` as a string with `useAsTitle: 'name'` on the related collection
- Array of `id`s
- Object that implements [`type ValueWithRelation`](https://github.com/payloadcms/payload/blob/934b443b5b3b0f610767786f940175b9db0b2da7/src/fields/config/types.ts#L283)
- Array of `ValueWithRelation` objects

**Comment by jmikrut** [comment]
Hey @evelynhathaway — this is all good stuff!

I agree with everything that you're saying. I believe the reason we restricted the `relationship` field to accept only a function for `defaultValue` is that generally you need to know the ID of the documents created before you can set the `defaultValue` accordingly, and the ID would typically be fetched as a result within a function, but I guess this doesn't necessarily need to be the case. We can probably remove that restriction.

In addition, I think that we can improve the types here so that the `defaultValue` type is inferred based on the field type.

Going to add this to our roadmap priority 2 and convert it to a discussion. This would also be a great PR if you are looking to contribute! Otherwise the Payload team will definitely address this.

Thank you!!

**Reply by evelynhathaway** [reply]
That sounds great!

> the ID would typically be fetched as a result within a function

How could that be done? It looks like `payload.find` would be a great solution, but the local API appears to be unavailable. I think an example would be great to document.

---

**`/src/collections/Posts.ts`**

```ts
import payload from 'payload';
import { CollectionConfig } from 'payload/types';

const Posts: CollectionConfig = {
  slug: 'posts',
  fields: [
    {
      name: 'category',
      type: 'relationship',
      relationTo: 'categories',
      defaultValue: async () => {
        const categoryDocuments = await payload.find({
          collection: 'categories',
          where: {
            name: {
              equals: "News",
            },
          },
        });
        return categoryDocuments.docs[0];
      },
    },
  ],
};

export default Posts;
```

**Error in DevTools**
```
Uncaught (in promise) TypeError: payload__WEBPACK_IMPORTED_MODULE_0__.default.find is not a function
```

**Snippet of `/src/collections/Posts.ts` in the Webpack Bundle**
```
var payload__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! payload */
	"./node_modules/payload/dist/webpack/mocks/emptyModule.js");

const Posts = {
// [...]
```

**Reply by jmikrut** [reply]
You'd need to use fetch, because the `defaultValue` function is run in the browser as well as the server. So `fetch` is the safest way for sure.

**Reply by evelynhathaway** [reply]
Oh right, of course! This worked for me.

```ts
      // [...]
      defaultValue: async () => {
        const response = await fetch("/api/categories?limit=1&where[name][equals]=News");
        const data: PaginatedDocs<{ id: string }> = await response.json();
        const [newsCategory] = data.docs;
        return newsCategory?.id;
      },
      // [...]
```

**Comment by johannesschaffer** [comment]
The documentation also doesn't explain how to set a defaultValue for an array

**Comment by ssyberg** [comment]
Oh man I just lost a few hours to this, let's update the docs!

**Comment by drecdroid** [comment]
I'm having problems with this:
```ts
{
      filterOptions: ({ data }) => {
        return {
          spaces: {
            contains: data?.space?.id
          }
        }
      },
      name: "author",
      type: "relationship",
      relationTo: "users",
      admin: {
        position: "sidebar",
        allowCreate: false
      },
      defaultValue: ({ user }) => [user.id],
      hasMany: true
    },
```

<img width="660" alt="image" src="https://github.com/payloadcms/payload/assets/922015/9c489f94-c288-4bca-959e-bea0be177ca9">

<img width="1035" alt="image" src="https://github.com/payloadcms/payload/assets/922015/a043c0c5-ebc4-4d84-8d35-7e8260d3f324">

**Comment by JavierBrooktec** [comment]
For anyone wondering how to pass default data to an array, debug it with the admin condition

```
admin: {
 condition: (data, siblingData, { user }) => {
  console.log(data);
  return true;
  },
},
```

and pass a similar structure to your default value
